Tests/Unit/Search/EventSubscriber/DoctrineOrmIndexRebuildSubscriberTest.php

```diff
@@ -0,0 +1,157 @@
+
+    /**
+     * @var SearchManager
+     */
+    private $searchManager;
+
+    /**
+     * @var DoctrineOrmIndexRebuildSubscriber
+     */
+    private $subscriber;
+
+
+    public function setUp()
+    {
+        $this->output = new BufferedOutput();
+        $this->objectManager = $this->prophesize(ObjectManager::class);
+        $this->ormMetadataFactory = $this->prophesize(ClassMetadataFactory::class);
+        $this->ormMetadata = $this->prophesize(OrmMetadata::class);
+        $this->searchMetadataFactory = $this->prophesize(MetadataFactory::class);
+        $this->searchHierarchy = $this->prophesize(ClassHierarchyMetadata::class);
+        $this->searchMetadata = $this->prophesize(ClassMetadata::class);
+        $this->searchManager = $this->prophesize(SearchManager::class);
+
+        $this->subscriber = new DoctrineOrmIndexRebuildSubscriber(
+            $this->objectManager->reveal(),
+            $this->searchMetadataFactory->reveal(),
+            $this->searchManager->reveal()
+        );
+
+    }
+
+    /**
+     * It should call the named repository method if requested
+     */
+    public function testRepositoryMethod()
+    {
+        $this->prepareReindex();
+
+        $this->searchMetadata->getReindexRepositoryMethod()->willReturn('foobar');
+        $this->searchManager->index(Argument::type('stdClass'))->shouldBeCalledTimes(1);
+
+        $event = $this->createEvent(null, false);
+        $this->subscriber->rebuildIndex($event);
+
+        $this->assertContains('1 entities indexed', $this->output->fetch());
+    }
+
+    /**
+     * It should throw an exception if the repository method does not exist
+     *
+     * @expectedException InvalidArgumentException
+     * @expectedExceptionMessage Repository method "barfoo" does not exist
+     */
+    public function testRepositoryMethodNotExists()
+    {
+        $this->prepareReindex();
+
+        $this->searchMetadata->getReindexRepositoryMethod()->willReturn('barfoo');
+
+        $event = $this->createEvent(null, false);
+        $this->subscriber->rebuildIndex($event);
+    }
+
+    private function createEvent($filter, $purge)
+    {
+        return new IndexRebuildEvent(
+            $filter,
+            $purge,
+            $this->output
+        );
+    }
+
+    private function prepareReindex()
+    {
+        $this->objectManager->getMetadataFactory()->willReturn($this->ormMetadataFactory);
+        $this->searchHierarchy->getOutsideClassMetadata()->willReturn($this->searchMetadata->reveal());
+        $this->ormMetadataFactory->getAllMetadata()->willReturn(array(
+            $this->ormMetadata->reveal()
+        ));
+        $this->ormMetadata->name = 'stdClass';
+        $this->searchMetadata->name = 'stdClass';
+        $this->searchMetadataFactory->getMetadataForClass('stdClass')->willReturn($this->searchHierarchy->reveal());
+
+        $this->objectManager->getRepository('stdClass')->willReturn(new TestRepository());
+    }
+}
+
+class TestRepository
```

Comment:
It needs concrete methods to test the Exception if the method does not
exist. Not sure if prophecy does that (tbh I didn't look).

On Wed, Oct 14, 2015 at 07:48:00AM -0700, Daniel Rotter wrote:
>    In
>    [1]Tests/Unit/Search/EventSubscriber/DoctrineOrmIndexRebuildSubscriberTest.php:
> 
>  > +    private function prepareReindex()
>  > +    {
>  > +        $this->objectManager->getMetadataFactory()->willReturn($this->ormMetadataFactory);
>  > +        $this->searchHierarchy->getOutsideClassMetadata()->willReturn($this->searchMetadata->reveal());
>  > +        $this->ormMetadataFactory->getAllMetadata()->willReturn(array(
>  > +            $this->ormMetadata->reveal()
>  > +        ));
>  > +        $this->ormMetadata->name = 'stdClass';
>  > +        $this->searchMetadata->name = 'stdClass';
>  > +        $this->searchMetadataFactory->getMetadataForClass('stdClass')->willReturn($this->searchHierarchy->reveal());
>  > +
>  > +        $this->objectManager->getRepository('stdClass')->willReturn(new TestRepository());
>  > +    }
>  > +}
>  > +
>  > +class TestRepository
> 
>    Why are you doing this? Why don't you use another mock for that?
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/massiveart/MassiveSearchBundle/pull/61#discussion_r42002975
>    2. https://github.com/massiveart/MassiveSearchBundle/pull/61/files#r42002975
>    3. https://github.com/massiveart/MassiveSearchBundle/pull/61/files#r42002975